8276559: (httpclient) Consider adding an HttpRequest.Builder.HEAD method to build a HEAD request.

jaikiran opened the pull request:
Can I please get a review for this change which  implements the enhancement noted in https://bugs.openjdk.java.net/browse/JDK-8276559?

The commit in this PR introduces a new `HEAD()` method on the `HttpRequest.Builder` interface. Given that this is a public interface on a public class of a public module, I decided to add this new method as a `default` method to prevent any potential breakages of any non-JDK implementations of this interface. The internal `jdk.internal.net.http.HttpRequestBuilderImpl` which implements this interface, overrides this default implementation to use an implementation of its own.

Existing tests have been updated to include coverage for this new method.

P.S: Slightly unrelated question - is it intentional that the `Builder` interface specifies the visibility modifiers on the interface methods. For example:  `public abstract Builder timeout(Duration duration);`, `public Builder headers(String... headers);` and so on?

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8276559](https://bugs.openjdk.java.net/browse/JDK-8276559): (httpclient) Consider adding an HttpRequest.Builder.HEAD method to build a HEAD request.


### Reviewers
 * [Christian Stein](https://openjdk.java.net/census#cstein) (@sormuras - Author) ⚠️ Review applies to 4ac690242996027a1a990563cd3efae470edffc0
 * [Daniel Fuchs](https://openjdk.java.net/census#dfuchs) (@dfuch - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/6348/head:pull/6348` \
`$ git checkout pull/6348`

Update a local copy of the PR: \
`$ git checkout pull/6348` \
`$ git pull https://git.openjdk.java.net/jdk pull/6348/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 6348`

View PR using the GUI difftool: \
`$ git pr show -t 6348`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/6348.diff">https://git.openjdk.java.net/jdk/pull/6348.diff</a>

</details>

jaikiran commented:
/csr

jaikiran commented:
I've opened a CSR request for this change https://bugs.openjdk.java.net/browse/JDK-8276996

sormuras reviewed (approved):
LGTM

dfuch commented on src/java.net.http/share/classes/java/net/http/HttpRequest.java:
```diff
@@ -366,6 +379,7 @@ public static Builder newBuilder(HttpRequest request, BiPredicate<String, String
                     switch (method) {
                         case "GET" -> builder.GET();
                         case "DELETE" -> builder.DELETE();
+                        case "HEAD" -> builder.HEAD();
```
The comment above (line 376)  mentions:
```
                // otherwise, the body is absent, special case for GET/DELETE,
                // or else use empty body
```
maybe HEAD should be added there as well...

jaikiran replied:
You're right. I missed that. I've updated the PR to fix this.

dfuch commented on test/jdk/java/net/httpclient/HttpRequestNewBuilderTest.java:
```diff
@@ -121,6 +121,7 @@ record NamedAssertion(String name, BiConsumer<HttpRequest, HttpRequest> test) {
                         .headers("testName1", "z").build() },
                 // dedicated method
                 { HttpRequest.newBuilder(URI.create("https://method-1/")).GET().build() },
+                { HttpRequest.newBuilder(URI.create("https://method-1/")).HEAD().build() },
```
I'd suggest to change the URI path to `/method-0` and move the HEAD case to just after line 122 (before GET)

jaikiran replied:
Done in the updated version of the PR

dfuch reviewed (commented):
Maybe `test/jdk/java/net/httpclient/HeadTest.java` could acquire a new dedicated method where `HttpRequest.Builder.HEAD()` is used. The test currently tests `method("HEAD", BodyPublishers.noBody());`

dfuch commented:
> P.S: Slightly unrelated question - is it intentional that the `Builder` interface specifies the visibility modifiers on the interface methods. For example: `public abstract Builder timeout(Duration duration);`, `public Builder headers(String... headers);` and so on?

Stylistic preference of the author - I guess (could be me). If this were a new class I would advise to remove the redundant qualifiers.

jaikiran commented:
> > P.S: Slightly unrelated question - is it intentional that the `Builder` interface specifies the visibility modifiers on the interface methods. For example: `public abstract Builder timeout(Duration duration);`, `public Builder headers(String... headers);` and so on?
> 
> Stylistic preference of the author - I guess (could be me). If this were a new class I would advise to remove the redundant qualifiers.

Understood. I have the habit of using the `public` too sometimes. It was the presence of `abstract` in one of the methods which made me a bit curious.

jaikiran commented:
> Maybe `test/jdk/java/net/httpclient/HeadTest.java` could acquire a new dedicated method where `HttpRequest.Builder.HEAD()` is used. The test currently tests `method("HEAD", BodyPublishers.noBody());`

I've updated the `HeadTest` to include request building with the convenience methods too. While I was at it, I noticed that test had unused imports and unused fields. I cleaned them up too in the updated PR. I hope that's OK.
Finally, I also noticed that I missed adding a `@bug` to these updated tests, so I went ahead and added them in my updated PR.

dfuch commented:
> > > P.S: Slightly unrelated question - is it intentional that the `Builder` interface specifies the visibility modifiers on the interface methods. For example: `public abstract Builder timeout(Duration duration);`, `public Builder headers(String... headers);` and so on?
> > 
> > 
> > Stylistic preference of the author - I guess (could be me). If this were a new class I would advise to remove the redundant qualifiers.
> 
> Understood. I have the habit of using the `public` too sometimes. It was the presence of `abstract` in one of the methods which made me a bit curious.

Oh - maybe it went back and forth from being an abstract class or an interface during prototyping...

jaikiran commented:
The CSR for this has been approved.

dfuch commented:
The changes look good. I'm going to give it a round of testing and I will approve if the results come clean.

dfuch reviewed (approved):
LGTM

jaikiran commented:
Thank you Daniel and @sormuras for the reviews.

jaikiran commented:
/integrate
